crates/html-macro/src/tag.rs

```diff
@@ -168,31 +221,46 @@ fn parse_text_node(input: &mut ParseStream) -> Result<Tag> {
 
     let mut idx = 0;
 
+    let mut start_span = None;
+
+    let mut most_recent_span: Option<Span> = None;
+
     loop {
         if input.is_empty() {
             break;
         }
 
-        // TODO: Ditch this and build text with proper spacing by using span start/end locations
-        if input.peek(Token![,]) {
-            let _: TokenTree = input.parse()?;
-            text += ",";
-        } else if input.peek(Token![!]) {
-            let _: TokenTree = input.parse()?;
-            text += "!";
-        } else if input.peek(Token![.]) {
-            let _: TokenTree = input.parse()?;
-            text += ".";
-        } else {
-            let tt: TokenTree = input.parse()?;
+        let tt: TokenTree = input.parse()?;
 
-            if idx != 0 {
-                text += " ";
-            }
+        if idx == 0 {
+            start_span = Some(tt.span());
+            most_recent_span = Some(tt.span());
+        }
 
-            text += &tt.to_string();
+        // Insert necessary space in between the tokens in this text node that we're building.
+        // In the real browser DOM there's no difference between one space and many,
+        // so we just insert one.
```

Comment:
That's not necessarily true! If you use `white-space: pre` or `white-space: pre-wrap` CSS rules then white-space will not be collapsed.

https://developer.mozilla.org/en-US/docs/Web/CSS/white-space

Not sure if that could be a problem or not.

Reply:
Ahhhh thanks for catching this!

Luckily the fix here is straightforwards enough.

We'll just look at the spans for the text tokens and see how far apart they are - then insert that much space in between then.

And if their spans are on different lines we'll just insert a new line character followed by the amount of spacing that.

---

Also in `text.rs` we'll need to insert the right amount of space and newlines before/after the entire string of text.. vs. right now we just insert a single space

https://github.com/chinedufn/percy/blob/edd4122ccee157566e5118e4c3e74d210eb3a091/crates/html-macro/src/parser/text.rs#L34-L36

---

Thanks for catching this - I'll add some more test cases and get this fixed. Then we should be in awesome shape here!

Reply:
Took a look and think there's a better solution.. Going to merge this and open a separate issue.

Reply:
#98